### DISCUSSION TITLE
Linking Pangeo Forge and Beam to AWS-S3

### DISCUSSION THREAD

**Discussion by mattjbr123** [NERC-CEH/rse_group#21] [General]
This might be better off as an issue, but: For FDRI, as part of the 'Gridded Datasets' product, we want to have a tool that can convert gridded datasets into an 'Analysis Ready Cloud Optimized' format, typically zarr, something that Pangeo Forge has the ability to do. 

The idea with Pangeo Forge is to provide a set of 'pipeline tasks' that would typically be used and adapted by users to convert their gridded data into ARCO (zarr). It used to use dask, but now solely uses apache beam as the backend, and writes the pipeline tasks in Beam's syntax. It's essentially a python package that is a wrapper around Beam, interfacing with the usual pangeo tools (xarray mainly). 
They used to provide the compute platform for this, but no longer, meaning that to use it properly we'd have to deploy Apache Beam somewhere, e.g. AWS or on-prem.

Beam uses backends called 'runners' to execute the workflows. Where these can be run is shown here: https://beam.apache.org/documentation/runners/capability-matrix/ 
You can get away with running Apache Beam's 'Direct Runner' on any linux machine/cluster/cloud compute service, as we did in [this demo](https://github.com/NERC-CEH/object_store_tutorial), but it's not as performant and can't write directly to object storage, along with other limitations, meaning it would be more fiddly to use to write data to the AWS-S3 store being developed as part of FDRI (or any other object storage solution we might want to use). 

The 'Flink' and 'Spark' runners could be run on AWS, which could link nicely with FDRI work if we use AWS going forward. However, this might be expensive. 
Alternative solutions could be: 
- to just run the direct runner locally and use a separate process to upload data to object storage once converted. But this is not ideal as it produces more complexity, and the direct runner has other issues that would make it less stable and reliable in an operational context.
- to find a way of deploying Flink or Spark or some other runner locally on-prem, on the private cloud or whatever, and using this to process the data and run our workflows. This is likely the cheaper option in terms of fees, but would need significant developer time and IT coordination which could take a while and turn out not to even be feasible. 
- Use something other than Beam. By replacing the Beam tools in pangeo-forge with those using a different workflow thingy (e.g. AirFlow?). If other products in FDRI use this it might save developer time as it's one less tool to get to grips with but it's also going to have an ongoing cost. 

This ultimately encapsulates a closed conversation I was having with Faiza that I thought others in the RSE group and possibly more widely would like to know about. To be updated as discussions progress...

**Comment by metazool** [comment]
I like Alternative 2, out of these. There's an on-prem Kubernetes cluster which could be used for prototyping, and an [Apache Flink helm chart](https://nightlies.apache.org/flink/flink-kubernetes-operator-docs-main/docs/operations/helm/) which could be run there without custom work. 

I had the impression comparing the original `pangeo-forge` paper [notes here](https://github.com/NERC-CEH/rse_group/blob/main/infrastructure/pipelines/pangeo_forge.md) to the current state of the examples, that the project has become less framework-neutral over time; from defaulting to [Prefect Flow](https://docs.prefect.io/2.6/concepts/flows/) to directly importing Beam dependencies in the onboarding examples. You could be storing up trouble by trying to consolidate on one view of a workflow-framework-thingy too soon; you've got space to experiment in right now.

How much hand-holding would we need from the Devops folks for permissions and bootstrap documentation for the self-managed k8s to deploy and test Flink? Maybe not much, what exists? I'd be up for a shared-screen session on this, ask someone from that team to join,

**Comment by mattjbr123** [comment]
Thanks @metazool - appreciate the valuable insight. 
For example I wasn't aware we had an on-prem kubernetes cluster... I'm still getting my head around how all these technologies link together - kubernetes, flink, spark, beam, beam's runners, helm, argo, airflow etc. - and how they are different and which is used for which bit of a workflow/pipeline. 

I started using pangeo-forge when it had just moved away from supporting dask and had tied itself to beam. I agree we shouldn't be tying ourselves to beam just because pangeo-forge has. If it's not too much work to instead use it as a template on which to implement whichever library (or whatever the appropriate word is) we go with or experiment with, then there's probably value to be gained from experimenting. 

I think as a starting point a discussion with devops about the potential for running flink would be useful. I will look to arrange...

**Reply by metazool** [reply]
I was hoping that if i linked an issue from another project into this Discussion it would show up here (as it does when you refer to one issue from another) but it doesn't appear to have.

Added a [reasonably generic summary of the workflow/pipeline stages](https://github.com/NERC-CEH/plankton_ml/issues/9) implicit in the plankton imaging project here, for reference. Happy to evaluate whatever components are most easily available against the use case...

**Comment by fsamreen** [comment]
The development team oversees a Kubernetes cluster for both staging and production environments. You can find detailed information about the cluster configuration, Helm repositories, and Rancher setup on our Wiki App Dev Site.

**FDRI** – the dev team has set up a Kubernetes cluster on AWS and is working on the implementation of the Time Series product, aiming to manage the complete end-to-end pipeline (ingestion, storage, QC, processing, publishing, metadata/provenance management) in a more automated manner. https://github.com/NERC-CEH/dri-infrastructure. @mattjbr123 We will discuss these aspects in detail as we work through the implementation of your product, covering infrastructure, technology, and skill requirements. We could look into K8 on AWS - Alternatively, we could consider having K8 deployed on the internal cloud to support the testbed deployments like Apache Beam - we can come back to this later as part of our long-term plans.

As part of our ongoing investigation to support the ingestion service, we are exploring several stable workflow technologies, including Argo Workflows, Apache Airflow, and Apache Kafka. These tools are primarily being evaluated for their capabilities to automate QCs at various data levels and layers, as well as to facilitate automated data ingestion from different sources in real-time or near real-time. We will document the output through ADRs to showcase our findings and inform our technological and architectural decisions. 
Apache Beam looks very interesting  - will be a new addition to the landscape as it offers a unified programming model for both batch and stream data processing, leveraging execution engines like Apache Flink and Spark.

There might not be a one-size-fits-all solution, so our investigation will involve testbed deployments and a deep dive into the functionalities of various tools/technologies. For instance, we are/will be evaluating solutions for low-latency, high-throughput data streaming, comparing execution engines to handle different scales and complexities of data processing. Our architectural focus is on cloud-native principles to build scalable, interoperable, and portable systems, utilising a multi-cloud approach - that is a space where we could get benefit from innovative technologies and solutions.

**Helm chart** is a package manager to simplify the process of deploying and managing applications on Kubernetes - easier to manage application updates, and rollbacks + support reusable and repeatable deployments. https://helm.sh/docs/intro/quickstart/

**Reply by mattjbr123** [reply]
Thank you @fsamreen for taking the time to explain these things, really really helpful!

**Comment by metazool** [comment]
Thank you for the detailed wide view @fsamreen !

This project https://github.com/NERC-CEH/dri-infrastructure 404s for me. If public work on it is either not being considered yet or waiting until it starts to look more complete, is it possible to add the EDS RSE group as members en masse? At @jmarshrossney's suggestion we set up a Github Team for that https://github.com/orgs/NERC-CEH/teams/eds-rse (so any permissions it gets, inherit to us all, and you can tag it in the same way you would an individual)

**Reply by dolegi** [reply]
I've granted read access to the eds-rse group. This is repo is private, write access to this repo grants you full access to AWS via a tool we are using https://www.runatlantis.io/ . Happy to discuss any bits you are curious about

**Comment by mattjbr123** [comment]
Have spoken to @dolegi. Off the back of this it looks like Argo and Airflow are not the right tools for this workflow. Remains to be seen if others exist that we could trial. @fsamreen [mentions](https://github.com/NERC-CEH/rse_group/discussions/21#discussioncomment-9990665) [Kafka](https://kafka.apache.org/), which I don't know anything about and might be worth a look. 

[Argo](https://argoproj.github.io/) and [Airflow](https://airflow.apache.org/) are more "workflow managers/schedulers/orchestrators", that just manage task dependencies in a workflow, whereas [Beam](https://beam.apache.org/), or more specifically the [Beam 'runners'](https://beam.apache.org/documentation/runners/capability-matrix/), is the thing that handles all the complexity of parallelising your code under the hood. If you write it in a certain style, Beam can read it and automatically determine the best way to parallelise it all for a particular system architecture under the hood. In Dask parlance it is generating the task graph, working out in which order and where to execute tasks, how to split them up, where to allocate memory etc...

So if we're looking at an alternative to Beam it would have to be something that does this, otherwise it will involve a fair amount of work as we'll have to manage at least some of this stuff ourselves. [Dask](https://www.dask.org/) is an option, but IMO Beam is a cut above. The problem with Dask is that how well the workflow runs is very dependent on how you write your code. You can essentially throw anything at Dask and it'll try to run it but it might not actually be any good, whereas Beam forces you to write a particular way and is a little more limited, but will still be performant. Though the performance will depend a little on which execution engine (or 'runner' in Beam parlance) is used, so there's still some work to be done there.

**Reply by dolegi** [reply]
Kafka is an event messaging system, my experience is using it as a beating heart of what is going on, where anybody can publish/subscribe, it's kinda like a queue on steroids, maybe not a great fit here either. 

Have you looked into using spark directly with pyspark?
We also discussed polars, which is simplier and limits to a single machine but might be okay for these workloads? It would just take longer, we can configure k8s to spin up a beefy instance.


If you want to setup a spark or flink cluster, we are using flux for both our AWS and on prem k8s cluster. AWS cluster is defined here https://github.com/NERC-CEH/dri-infrastructure-k8s-staging/tree/main, happy to help with this.

**Reply by mattjbr123** [reply]
I did have a look at [polars](https://pola.rs/), but looks like it is limited to data that fits well into tables (i.e. it reads like a souped up version of pandas), there's no mention of any support for gridded datasets like HDF/NetCDF/zarr in the [IO page](https://docs.pola.rs/user-guide/io/) or anywhere else obvious.

Using spark directly you say... I'm only just becoming aware that these '[runners](https://beam.apache.org/documentation/runners/capability-matrix/)' that Beam uses to execute workflows are actually workflow engines of a sort in themselves. Beam is a layer on top of them that translates what you want to do into a 'task graph' or 'execution plan' that is performant for the particular 'runner'/workflow engine you are using. So the short answer is no I haven't, but I am curious to look into this.

Beam can use Dask, Flink or Spark. I see a plan forming where we trial each of these, perhaps additionally without Beam on top too if we're feeling adventurous... Though the whole point of Beam is that it makes your workflow performant in whatever workflow engine you use, so why would we want to do things without it? Only if we think we can beat it I suppose.

**Reply by dolegi** [reply]
```python
from pyspark.sql import SparkSession
import xarray as xr
import numpy as np
from glob import glob
import os

input_dir = "input"
output_dir = "output"
    
# Initialize Spark session
spark = SparkSession.builder \
    .appName("NetCDF to Zarr Converter") \
    .getOrCreate()

# Function to process a single NetCDF file
def process_netcdf(file_path):
    ds = xr.open_dataset(file_path)
    
    # Do things with it here

    return ds

def convert_netcdf_to_zarr():
    # Get list of NetCDF files
    netcdf_files = glob(os.path.join(input_dir, "*.nc"))
    
    # Create Reso;oemt distributed dataset of file paths
    rdd = spark.sparkContext.parallelize(netcdf_files)
    
    # Process NetCDF files in parallel
    processed_rdd = rdd.map(process_netcdf)
    
    # Collect and combine
    datasets = processed_rdd.collect()
    combined_ds = xr.concat(datasets, dim="time")
    
    # Save as Zarr
    combined_ds.to_zarr(output_dir)

convert_netcdf_to_zarr()

spark.stop()
```

I had a quick play with pyspark and can convert a sample GEAR netcdf to zarr, I'm still getting my head around this but it's concatting on the time dimensions which probably isn't what we want.

We need to think about where we will be running this "pipeline" too, from speaking to @iwalmsley  about the wf datasets which convert ~70TB of netcdf to zarr, they need to use JASMIN, which limits us to dask, for the wf datasets it takes ~24hrs, using loads of nodes. Reading about beam on dask and it doesn't have great support ["The Dask runner remains relatively immature and untested; there are important beam features that are still unimplemented"](https://dask.discourse.group/t/distributed-runner-for-beam/784/4)

I think we can catch up today, commenting here to keep the thread up to date, love this knowledge sharing!

**Reply by mattjbr123** [reply]
Awesome, having that boilerplate simple spark code will be really useful, saves me having to get my head around it from scratch! It certainly looks promising, and definitely something to weigh up against any Beam implementations. Concatting in time is what we want with this dataset so no problems there. 

Yeah doesn't surprise me, the dask runner for Beam is a fairly new venture, probably one to avoid for now. Could you share @iwalmsley 's code if that's possible? We could learn something from that. When we last talked about, probably over a year ago now, it was using dask directly on JASMIN I think.

**Reply by mattjbr123** [reply]
[Project board for development work is here](https://github.com/orgs/NERC-CEH/projects/13/views/1) for those who'd like a closer look at things!

**Comment by mattjbr123** [comment]
I found this to be a very interesting thread over on the pangeo forums: https://discourse.pangeo.io/t/what-should-a-pangeo-2-0-cloud-tech-stack-look-like/4528 
@metazool something is making me feel like this might be of interest? It's an interesting discussion around on 'where do we go from here?' for pangeo, with a focus on managing/deploying the distributed systems that the pangeo stack increasinly embraces and runs on. 

It's also brought to my attention:
- this interesting attempt at [environment synchronisation](https://docs.coiled.io/user_guide/why.html#how-environment-synchronization) that the [coiled](https://docs.coiled.io/user_guide/index.html) platform uses, that might feed in to discussions about best-practice around python environments
- A [package](https://github.com/google/xarray-beam/tree/main) that integrates xarray directly with Beam somehow, which I think is an option we could explore for the [FDRI Gridded Data work](https://github.com/orgs/NERC-CEH/projects/13/views/1) if we want to be more explicit about the parallelisation of datasets. Only glanced at it so far
-
